x/orderbook/msg_test.go

```diff
@@ -0,0 +1,224 @@
+	}{
+		"success": {
+			msg: &CancelOrderMsg{
+				Metadata: &weave.Metadata{Schema: 1},
+				OrderID:  weavetest.SequenceID(5),
+			},
+			wantErr: nil,
+		},
+		"missing metadata": {
+			msg: &CancelOrderMsg{
+				OrderID: weavetest.SequenceID(5),
+			},
+			wantErr: errors.ErrMetadata,
+		},
+		"bad order id": {
+			msg: &CancelOrderMsg{
+				Metadata: &weave.Metadata{Schema: 1},
+				OrderID:  []byte{0, 0, 0, 0, 0, 1},
+			},
+			wantErr: errors.ErrInput,
+		},
+		"missing order id": {
+			msg: &CancelOrderMsg{
+				Metadata: &weave.Metadata{Schema: 1},
+			},
+			wantErr: errors.ErrEmpty,
+		},
+	}
+
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			if err := tc.msg.Validate(); !tc.wantErr.Is(err) {
+				t.Fatalf("unexpected error: %+v", err)
+			}
+		})
+	}
+}
+
+func TestValidateCreateOrderMsg(t *testing.T) {
+	trader := weavetest.NewCondition().Address()
+
+	cases := map[string]struct {
+		msg     weave.Msg
+		wantErr *errors.Error
+	}{
+		"success": {
+			msg: &CreateOrderMsg{
+				Metadata:    &weave.Metadata{Schema: 1},
+				Trader:      trader,
+				OrderBookID: weavetest.SequenceID(12345),
+				Offer:       coin.NewCoinp(100, 12345, "ETH"),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: nil,
+		},
+		"missing metadata": {
+			msg: &CreateOrderMsg{
+				Trader:      trader,
+				OrderBookID: weavetest.SequenceID(12345),
+				Offer:       coin.NewCoinp(100, 12345, "ETH"),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: errors.ErrMetadata,
+		},
+		"bad trader": {
+			msg: &CreateOrderMsg{
+				Metadata:    &weave.Metadata{Schema: 1},
+				Trader:      []byte{17, 29, 0, 8},
+				OrderBookID: weavetest.SequenceID(12345),
+				Offer:       coin.NewCoinp(100, 12345, "ETH"),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: errors.ErrInput,
+		},
+		"bad orderbook id": {
+			msg: &CreateOrderMsg{
+				Metadata:    &weave.Metadata{Schema: 1},
+				Trader:      trader,
+				OrderBookID: []byte{32, 0, 0, 1},
+				Offer:       coin.NewCoinp(100, 12345, "ETH"),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: errors.ErrInput,
+		},
+		"missing offer": {
+			msg: &CreateOrderMsg{
+				Metadata:    &weave.Metadata{Schema: 1},
+				Trader:      trader,
+				OrderBookID: weavetest.SequenceID(12345),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"invalid offer": {
+			msg: &CreateOrderMsg{
+				Metadata:    &weave.Metadata{Schema: 1},
+				Trader:      trader,
+				OrderBookID: weavetest.SequenceID(12345),
+				Offer:       coin.NewCoinp(1, 0, "1WIN!"),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: errors.ErrCurrency,
+		},
+		"negative offer": {
+			msg: &CreateOrderMsg{
+				Metadata:    &weave.Metadata{Schema: 1},
+				Trader:      trader,
+				OrderBookID: weavetest.SequenceID(12345),
+				Offer:       coin.NewCoinp(-5, 0, "ETH"),
+				Price:       NewAmountp(11, 0),
+			},
+			wantErr: errors.ErrInput,
+		},
+		// TODO: bad price
```

Comment:
May I ask why did you leave this test for later in particular?

Reply:
Because it requires implementing `Amount` logic, which I wanted to do in a future PR. There are a number of methods we need on Amount, and I'd rather group them, then update the tests/validation code that depends on it.

Validation and models seemed the first place to start. Then logic, then handlers. Then higher-level integrations.